File: src/main/java/net/clgd/ccemux/emulation/CCEmuXAPI.java
package net.clgd.ccemux.emulation;

import java.awt.Desktop;
import java.awt.Toolkit;
import java.awt.datatransfer.Clipboard;
import java.awt.datatransfer.StringSelection;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.Map;

import dan200.computercraft.api.peripheral.IPeripheral;
import dan200.computercraft.core.computer.Computer;
import net.clgd.ccemux.peripherals.PeripheralFactory;
import org.luaj.vm2.Lua;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import dan200.computercraft.api.lua.ILuaContext;
import dan200.computercraft.api.lua.LuaException;
import dan200.computercraft.core.apis.ILuaAPI;

@FunctionalInterface
interface APIMethod {
	Object[] accept(Object[] o) throws LuaException;
}

public class CCEmuXAPI implements ILuaAPI {
	private static final Logger log = LoggerFactory.getLogger(CCEmuXAPI.class);

	private final String name;
	private final Map<String, APIMethod> methods = new LinkedHashMap<>();

	public CCEmuXAPI(CCEmuX emu, EmulatedComputer computer, String name) {
		this.name = name;

		methods.put("getVersion", o -> new Object[] { CCEmuX.getVersion() });

		// TODO
		// methods.put("setCursorChar", o -> {
		// if (o.length < 1 || !(o[0] instanceof String)) {
		// throw new LuaException("expected string for argument #1");
		// }
		//
		// String s = (String) o[0];
		// if (s.length() < 1) {
		// throw new LuaException("cursor char cannot be empty");
		// }
		//
		// computer.cursorChar = s.charAt(0);
		//
		// return new Object[] {};
		// });

		methods.put("closeEmu", o -> {
			computer.shutdown();
			emu.removeComputer(computer);
			return new Object[] {};
		});

		methods.put("openEmu", o -> {
			int id = -1;

			if (o.length > 0 && o[0] != null) {
				if (o[0] instanceof Number) {
					id = ((Number) o[0]).intValue();
				} else {
					throw new LuaException("expected number or nil for argument #1");
				}
			}

			EmulatedComputer ec = emu.addComputer(id);

			return new Object[] { ec.getID() };
		});

		methods.put("openDataDir", o -> {
			try {
				Desktop.getDesktop().browse(emu.cfg.getDataDir().toUri());
				return new Object[] { true };
			} catch (Exception e) {
				return new Object[] { false, e.toString() };
			}
		});

		methods.put("milliTime", o -> {
			return new Object[] { System.currentTimeMillis() };
		});

		methods.put("nanoTime", o -> {
			return new Object[] { System.nanoTime() };
		});

		methods.put("echo", o -> {
			if (o.length > 0 && o[0] instanceof String) {
				log.info("[Computer {}] {}", computer.getID(), (String) o[0]);
			} else {
				throw new LuaException("expected string for argument #1");
			}

			return new Object[] {};
		});

		methods.put("setClipboard", o -> {
			if (o.length > 0 && o[0] instanceof String) {
				StringSelection sel = new StringSelection((String) o[0]);
				Clipboard c = Toolkit.getDefaultToolkit().getSystemClipboard();
				c.setContents(sel, sel);
			} else {
				throw new LuaException("expected string for argument #1");
			}

			return new Object[] {};
		});

		methods.put("attach", o -> {
			if (o.length > 1 && o[0] instanceof String && o[1] instanceof String) {
				String peripheralName = (String) o[0];
				String sideName = (String) o[1];
				int sideId = parseSide(sideName);
				IPeripheral peripheral = PeripheralFactory.implementations.get(peripheralName);

				if(sideId > 0) {
					if(peripheral != null) {
						computer.setPeripheral(sideId, peripheral);
					} else {
						throw new LuaException("invalid peripheral");
					}
				} else {
					throw new LuaException("invalid side");
				}
			} else if (o.length > 1 && !(o[1] instanceof String)) {
				throw new LuaException("expected string for argument #2");
			} else {
				throw new LuaException("expected string for argument #1");
			}

			return new Object[] {};
		});

		methods.put("detach", o -> {
			if (o.length > 0 && o[0] instanceof String) {
				String sideName = (String) o[0];
				int sideId = parseSide(sideName);

				if (sideId > 0) {
					computer.setPeripheral(sideId, null);
				} else {
					throw new LuaException("invalid side");
				}
			} else {
				throw new LuaException("expected string for argument #1");
			}

			return new Object[] {};
		});
	}

	private int parseSide(String side) {
		for(int n = 0; n < Computer.s_sideNames.length; ++n) {
			if(side.equals(Computer.s_sideNames[n])) {
				return n;
			}
		}

		return -1;
	}

	@Override
	public String[] getMethodNames() {
		return methods.keySet().toArray(new String[] {});
	}

	@Override
	public Object[] callMethod(ILuaContext context, int method, Object[] arguments)
			throws LuaException, InterruptedException {
		return new ArrayList<APIMethod>(methods.values()).get(method).accept(arguments);
	}

	@Override
	public void advance(double dt) {}

	@Override
	public String[] getNames() {
		return new String[] { name };
	}

	@Override
	public void shutdown() {}

	@Override
	public void startup() {}

}

File: src/main/kotlin/net/clgd/ccemux/emulation/CCEmuXAPI.kt
package net.clgd.ccemux.emulation

import dan200.computercraft.api.lua.ILuaContext
import dan200.computercraft.api.lua.LuaException
import dan200.computercraft.core.apis.ILuaAPI
import net.clgd.ccemux.rendering.RenderingMethod
import java.awt.Desktop
import java.util.*

class CCEmuXAPI(val computer: EmulatedComputer, val name: String) : ILuaAPI {
	// somewhat hacky but type inference doesn't cut it here
	val methods = LinkedHashMap<String, (Array<out Any>) -> Array<out Any>>()

	init {
		methods.put("getVersion") {
			arrayOf(CCEmuXConsts.getVersion())
		}

		methods.put("setCursorChar") {
			arr -> run {
				if (arr.isEmpty() || !(arr[0] is String)) {
					throw LuaException("expected string for argument #1")
				}

				val cursorChar = arr[0] as String

				if (cursorChar.isNotEmpty()) {
					computer.cursorChar = cursorChar.getOrElse(0, { i -> '?' })
				} else {
					throw LuaException("cursor char can't be empty")
				}

				emptyArray()
			}
		}

		methods.put("setResolution") {
			arr -> run {
				if (arr.size < 2 || !(arr.get(0) is java.lang.Number) || !(arr.get(1) is java.lang.Number)) {
					throw LuaException("expected number for arguments #1 and #2")
				}

				val width = (arr.get(0) as java.lang.Number).intValue()
				val height = (arr.get(1) as java.lang.Number).intValue()

				if (width <= 0 || height <= 0) {
					throw LuaException ("width and height must be above 0")
				}

				computer.resize(width, height)

				emptyArray()
			}
		}

		methods.put("saveSettings") {
			computer.emu.conf.saveProperties()
			emptyArray()
		}

		methods.put("closeEmu") {
			computer.dispose()
			emptyArray()
		}

		methods.put("openEmu") {
			arr -> run {
				var id = -1

				if (arr.isNotEmpty()) {
					val nid = arr[0]

					if (nid is java.lang.Number) {
						id = nid.intValue()
					} else {
						throw LuaException("expected number or nil for argument #1")
					}
				}

				val ec = computer.emu.createEmulatedComputer(id)
				val r = ec.emu.conf.getRenderer().map { r -> RenderingMethod.create(r, ec.emu, ec) }
				r.forEach({ b -> b?.setVisible(true) })

				arrayOf(ec.getID())
			}
		}

		methods.put("openDataDir") {
			try {
				Desktop.getDesktop().browse(computer.emu.dataDir.toUri())
				arrayOf(true)
			} catch (e: Exception) {
				arrayOf(false)
			}
		}

		methods.put("milliTime") { arrayOf(System.currentTimeMillis()) }
		methods.put("nanoTime") { arrayOf(System.nanoTime()) }

		methods.put("echo") {
			arr -> run {
				if (arr.isNotEmpty() && arr[0] is String) {
					computer.emu.logger.info("[Computer {}] {}", computer.getID(), arr[0] as String)
				} else {
					throw LuaException("expected string for argument #1")
				}

				emptyArray()
			}
		}
	}

	override fun advance(d: Double) {}

	override fun shutdown() {}

	override fun startup() {}

	override fun getNames() = arrayOf(name)

	override fun callMethod(lua: ILuaContext?, method: Int, arguments: Array<out Any>?): Array<out Any> {
		if (arguments != null) {
			return methods.getOrDefault(null, { i -> throw LuaException("this shouldn't happen!") })(arguments)
		} else {
			return arrayOf()
		}
	}

	override fun getMethodNames() = methods.keys.toTypedArray()
}
